test/test_driver_flags.py

```diff
@@ -0,0 +1,168 @@
+"""Tests auto_flush and read_only flags are use correctly when applied to psqlgraph
+
+   auto_flush and read_only flags can be applied via the PsqlGraphDriver constructor or when starting new
+   transactions using ` with session_scope()`
+
+   flags applied to the constructor can be overridden by applying the same flag while starting a transaction.
+   Once a flag is applied to an active session, all child sessions will use the same flag values specified in the
+   parent no matter what was passed in.
+"""
+
+import pytest
+from sqlalchemy import exc
+
+import psqlgraph
+from test import models
+
+
+@pytest.fixture(scope="module")
+def pg_no_auto_flush(pg_conf, pg_driver):
+    return psqlgraph.PsqlGraphDriver(auto_flush=False, **pg_conf)
+
+
+@pytest.fixture(scope="module")
+def pg_read_only(pg_conf, pg_driver):
+    return psqlgraph.PsqlGraphDriver(read_only=True, **pg_conf)
+
+
+def test_no_auto_flush_driver(pg_no_auto_flush):
+    """Tests auto flush settings are applied appropriately"""
+
+    # Uses value set by constructor
+    with pg_no_auto_flush.session_scope() as s:
+        assert s.autoflush is False
+
+    # Nested still uses value set by constructor
+    with pg_no_auto_flush.session_scope() as s:
+        with pg_no_auto_flush.session_scope(can_inherit=False):
+            assert s.autoflush is False
+
+    # override default value
+    with pg_no_auto_flush.session_scope(auto_flush=True) as s:
+        assert s.autoflush is True
+
+
+def test_auto_flush_driver(pg_driver):
+    """Tests default auto flush settings"""
+
+    # uses default constructor value
+    with pg_driver.session_scope() as s:
+        assert s.autoflush is True
+
+    # override constructor value to True
+    with pg_driver.session_scope(auto_flush=True) as s:
+        assert s.autoflush is True
+
+    # override constructor value to False
+    with pg_driver.session_scope(auto_flush=False) as s:
+        assert s.autoflush is False
+
+
+def test_params__nested_auto_flush(pg_driver):
+    """Tests inherited sessions cannot be changed"""
+
+    with pg_driver.session_scope() as s:
+        assert s.autoflush is True
+
+        # not changing auto_flush for an inherited transaction
+        with pg_driver.session_scope(auto_flush=False) as t:
```

Comment:
Should we throw exception when user try to change inherited session?

Reply:
Logging a warning makes sense, also an sql exception will be raised whenever one tries to write with a read only transaction, even when they tried changing it.